### PR TITLE
[Merged by Bors] - feat: prove Fermat's Last Theorem for `n=3`

### PR DESCRIPTION
**Pull request by riccardobrasca**
We add `fermatLastTheoremThree`: Fermat's Last Theorem for `n=3`.

From the flt3 project at LFTCM2024.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/FLT/Three.lean [hunk 1]
```diff
@@ -584,10 +576,168 @@ private noncomputable def u₃ :=(exists_cube_associated S).2.2.choose_spec.choo
...
+/-- Given `S : Solution`, we have that
+`S.Y ^ 3 + (S.u₄ * S.Z) ^ 3 = S.u₅ * (λ ^ (S.multiplicity - 1) * S.X) ^ 3`. -/
+private lemma formula3 :
+    S.Y ^ 3 + (S.u₄ * S.Z) ^ 3 = S.u₅ * (λ ^ (S.multiplicity - 1) * S.X) ^ 3 := by
+  calc S.Y^3+(S.u₄*S.Z)^3=S.Y^3+S.u₄^2*S.u₄*S.Z^3 := by ring
+  _ = S.Y^3+S.u₄*S.Z^3 := by simp [← Units.val_pow_eq_pow_val, S.u₄_sq]
+  _ = S.u₅*(λ^(S.multiplicity-1)*S.X)^3 := S.formula2
+
+/-- Given `S : Solution`, we construct `S₁ : Solution'`, with smaller multiplicity of `λ` in
+  `c` (see ``). -/
```
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/FLT/Three.lean:674]: See what?

File: Mathlib/NumberTheory/FLT/Three.lean [hunk 2]
```diff
@@ -584,10 +576,168 @@ private noncomputable def u₃ :=(exists_cube_associated S).2.2.choose_spec.choo
...
+/-- We have that `S.Solution'_descent.multiplicity < S.multiplicity`. -/
+lemma Solution'_descent_multiplicity_lt :
+    (Solution'_descent S).multiplicity < S.multiplicity := by
+  rw [Solution'_descent_multiplicity S, Nat.sub_one]
+  exact Nat.pred_lt <| by have := S.two_le_multiplicity; omega
+
+/-- Given any `S : Solution`, there is another `S₁ : Solution` such that
+  `S₁.multiplicity < S.multiplicity` -/
+theorem exists_Solution_multiplicity_lt :
+    ∃ (S₁ : Solution hζ), S₁.multiplicity < S.multiplicity := by
```
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/FLT/Three.lean:708]:
```suggestion
    ∃ S₁ : Solution hζ, S₁.multiplicity < S.multiplicity := by
```

File: Mathlib/NumberTheory/FLT/Three.lean [hunk 3]
```diff
@@ -584,10 +576,168 @@ private noncomputable def u₃ :=(exists_cube_associated S).2.2.choose_spec.choo
...
 
 end FermatLastTheoremForThreeGen
 
 end eisenstein
 
 end case2
+
+/-- Fermat's Last Theorem for `n = 3`: if `a b c : ℕ` are all non-zero then
+`a ^ 3 + b ^ 3 ≠ c ^ 3`. -/
+theorem fermatLastTheoremThree : FermatLastTheoremFor 3 := by
```
Comment (Ruben-VandeVelde) [inline Mathlib/NumberTheory/FLT/Three.lean:722]: :tada:

**Review by Ruben-VandeVelde — APPROVED** [review]

File: Mathlib/NumberTheory/FLT/Three.lean [hunk 4]
```diff
@@ -11,33 +11,40 @@ import Mathlib.Algebra.Ring.Divisibility.Lemmas
 
 /-!
 # Fermat Last Theorem in the case `n = 3`
-The goal of this file is to prove Fermat Last theorem in the case `n = 3`.
+The goal of this file is to prove Fermat Last Theorem in the case `n = 3`.
```
Comment (semorrison) [inline Mathlib/NumberTheory/FLT/Three.lean:14]:
```suggestion
The goal of this file is to prove Fermat's Last Theorem in the case `n = 3`.
```

**Comment by semorrison** [comment]
bors d+

**Pull request closed without merging**